ci/precise_benchmarks.py

```diff
@@ -0,0 +1,150 @@
+
+def split_commas_into_token(string):
+    return [str_.strip() for str_ in string.split(',')]
+def split_space_into_token(string):
+    return [str_.strip() for str_ in string.split(' ')]
+ 
+def main():
+    code_cache, out_cache, err = run(["valgrind", "--tool=cachegrind", sys.argv[1]])
+    cachegrind_results = generic_processor(code_cache, out_cache.decode('utf-8'))
+    code_mem, out_mem, err = run(["valgrind", "--tool=memcheck", sys.argv[1]])
+    memcheck_results = generic_processor(code_mem, out_mem.decode('utf-8'))
+    code_instr, out_instr, err = run(["valgrind", "--tool=callgrind", sys.argv[1]])
+    instr_results = generic_processor(code_instr, out_instr.decode('utf-8'))
+    repo = git.Repo(search_parent_directories=True)
+    useful_bench_info = {
+            "Operating System": platform.platform(),
+            "Platform Uname": platform.uname(),
+            "Binary Size": int(os.path.getsize(sys.argv[1])),
+            "Git Commit": repo.head.object.hexsha,
+    }
+    cachegrind_result_dict = {
+            "I_ref": int(remove_comma(cachegrind_results[0].strip())),
+            "I1_misses": int(remove_comma(cachegrind_results[1].strip())),
+            "LLi_misses": int(remove_comma(cachegrind_results[2].strip())),
+            "I1_miss_rate": percent_to_float(float(remove_percent(cachegrind_results[3].strip()))),
+            "LLi_miss_rate": percent_to_float(float(remove_percent(remove_percent(cachegrind_results[4].strip())))),
+            "D_ref": int(remove_comma(cachegrind_results[5].strip())),
+            "D1_misses": int(remove_comma(cachegrind_results[6].strip())),
+            "LLd_misses": int(remove_comma(cachegrind_results[7].strip())),
+            "D1_miss_rate": percent_to_float(float(remove_percent(cachegrind_results[8]))),
+            "LLd_miss_rate": percent_to_float(float(remove_percent(cachegrind_results[9].strip()))),
+            "LL_ref": int(remove_comma(cachegrind_results[10].strip())),
+            "LL_misses": int(remove_comma(cachegrind_results[11].strip())),
+            "LL_miss_rate": percent_to_float(float(remove_percent(cachegrind_results[12].strip()))),
+    }
+    memcheck_result_dict = {
+            "HEAP total usage at exit": int(split_space_into_token(remove_comma(memcheck_results[1]))[0]),
+            "HEAP total usage": int(split_space_into_token(remove_comma(memcheck_results[2]))[4])
+    }
+    instruction_result_dict = {
+            "Instruction reads": int(remove_comma(instr_results[2]))
+    }
+    metrics_dict = cachegrind_result_dict;
+    metrics_dict.update(memcheck_result_dict)
+    metrics_dict.update(instruction_result_dict)
+    final_dict = {
+            "System Information": useful_bench_info,
+            "Metrics": metrics_dict
+    }
+    output_file_name = strftime("%Y-%m-%d %H:%M:%S", gmtime())
+    create_json_from_dict(final_dict, output_file_name)
+    
+if __name__ == "__main__":
+    main()
+    os.system("rm -f cachegrind.out* callgrind.out*")
```

Comment:
Blocking: Don't delete files you didn't create! If the user happened to have some cachegrind or callgrind files, this line would delete them. Naughty!